File: pixie/cli/dataset_command.py
"""``pixie dataset`` CLI commands.

Provides two async operations for saving trace data from
:class:`~pixie.storage.store.ObservationStore` into a
:class:`~pixie.dataset.store.DatasetStore`:

- :func:`dataset_create` — create a new dataset from a trace's root span.
- :func:`dataset_append` — append a trace's root span to an existing dataset.
"""

from __future__ import annotations

from pixie.dataset.models import Dataset
from pixie.dataset.store import DatasetStore
from pixie.storage.evaluable import as_evaluable
from pixie.storage.store import ObservationStore


async def dataset_create(
    *,
    name: str,
    trace_id: str,
    observation_store: ObservationStore,
    dataset_store: DatasetStore,
) -> Dataset:
    """Create a new dataset containing the root span of a trace as an evaluable item.

    Args:
        name: Unique name for the new dataset.
        trace_id: Trace ID whose root span will be converted to an evaluable item.
        observation_store: Store to read spans from.
        dataset_store: Store to write the dataset to.

    Returns:
        The created ``Dataset``.

    Raises:
        ValueError: If no root observation exists for *trace_id*.
        FileExistsError: If a dataset with *name* already exists.
    """
    root = await observation_store.get_root(trace_id)
    evaluable = as_evaluable(root)
    return dataset_store.create(name, items=[evaluable])


async def dataset_append(
    *,
    name: str,
    trace_id: str,
    observation_store: ObservationStore,
    dataset_store: DatasetStore,
) -> Dataset:
    """Append the root span of a trace as an evaluable item to an existing dataset.

    Args:
        name: Name of the existing dataset.
        trace_id: Trace ID whose root span will be converted to an evaluable item.
        observation_store: Store to read spans from.
        dataset_store: Store to write the updated dataset to.

    Returns:
        The updated ``Dataset``.

    Raises:
        ValueError: If no root observation exists for *trace_id*.
        FileNotFoundError: If no dataset with *name* exists.
    """
    root = await observation_store.get_root(trace_id)
    evaluable = as_evaluable(root)
    return dataset_store.append(name, evaluable)

File: ts/src/cli/datasetCommand.ts
/**
 * `pixie-qa dataset` CLI commands.
 *
 * Provides operations for managing datasets and saving trace spans
 * as evaluable items:
 *
 * - `datasetCreate` — create a new empty dataset.
 * - `datasetList` — list datasets with basic information.
 * - `datasetSave` — select a span from the latest trace and save it.
 */

import type { Dataset } from "../dataset/models";
import { DatasetStore } from "../dataset/store";
import type { Evaluable, JsonValue } from "../storage/evaluable";
import { UNSET, asEvaluable } from "../storage/evaluable";
import type { Unset } from "../storage/evaluable";
import { ObservationStore } from "../storage/store";
import { getConfig } from "../config";

/**
 * Create a new empty dataset.
 */
export function datasetCreate(
  name: string,
  datasetStore?: DatasetStore
): Dataset {
  const store = datasetStore ?? new DatasetStore();
  return store.create(name);
}

/**
 * Return metadata for every dataset.
 */
export function datasetList(
  datasetStore?: DatasetStore
): Array<Record<string, unknown>> {
  const store = datasetStore ?? new DatasetStore();
  return store.listDetails();
}

/**
 * Format dataset metadata rows as an aligned CLI table.
 */
export function formatDatasetTable(
  rows: Array<Record<string, unknown>>
): string {
  if (rows.length === 0) return "No datasets found.";

  const headers = ["Name", "Rows", "Created", "Updated"];
  const data = rows.map((r) => [
    String(r["name"]),
    String(r["rowCount"] ?? r["row_count"] ?? 0),
    String(r["createdAt"] ?? r["created_at"] ?? ""),
    String(r["updatedAt"] ?? r["updated_at"] ?? ""),
  ]);

  const colWidths = headers.map((h) => h.length);
  for (const row of data) {
    for (let i = 0; i < row.length; i++) {
      colWidths[i] = Math.max(colWidths[i], row[i].length);
    }
  }

  const fmtRow = (cells: string[]): string =>
    cells.map((c, i) => c.padEnd(colWidths[i])).join("  ");

  const lines = [
    fmtRow(headers),
    fmtRow(colWidths.map((w) => "-".repeat(w))),
    ...data.map(fmtRow),
  ];
  return lines.join("\n");
}

/**
 * Select a span from the latest trace and save it to a dataset.
 */
export function datasetSave(opts: {
  name: string;
  select?: string;
  spanName?: string;
  expectedOutput?: JsonValue | Unset;
  notes?: string;
}): Dataset {
  const config = getConfig();
  const obsStore = new ObservationStore(config.dbPath);
  const dsStore = new DatasetStore();

  try {
    obsStore.createTables();
    const traces = obsStore.listTraces(1);
    if (traces.length === 0) {
      throw new Error("No traces found in the observation store.");
    }
    const traceId = traces[0]["traceId"] as string;

    const select = opts.select ?? "root";
    let span;

    if (select === "root") {
      span = obsStore.getRoot(traceId);
    } else if (select === "last_llm_call") {
      span = obsStore.getLastLlm(traceId);
      if (span === null) {
        throw new Error(`No LLM span found in trace ${traceId}.`);
      }
    } else if (select === "by_name") {
      if (!opts.spanName) {
        throw new Error("--span-name is required when selection mode is 'by_name'.");
      }
      const matches = obsStore.getByName(opts.spanName, traceId);
      if (matches.length === 0) {
        throw new Error(
          `No span named '${opts.spanName}' found in trace ${traceId}.`
        );
      }
      span = matches[matches.length - 1];
    } else {
      throw new Error(`Unknown selection mode: '${select}'`);
    }

    let evaluable: Evaluable = asEvaluable(span);

    // Apply expected_output if provided
    const expectedOutput = opts.expectedOutput ?? UNSET;
    if (expectedOutput !== UNSET) {
      evaluable = {
        ...evaluable,
        expectedOutput: expectedOutput as JsonValue,
      };
    }

    // Apply notes if provided
    if (opts.notes !== undefined) {
      const existingMeta = evaluable.evalMetadata
        ? { ...evaluable.evalMetadata }
        : {};
      existingMeta["notes"] = opts.notes;
      evaluable = {
        ...evaluable,
        evalMetadata: existingMeta,
      };
    }

    return dsStore.append(opts.name, evaluable);
  } finally {
    obsStore.close();
  }
}
